src/transaction_builder.js

```diff
@@ -219,19 +223,99 @@ export class TransactionBuilder {
 
     attrs.timeBounds = new xdr.TimeBounds(this.timebounds);
 
-    const xtx = new xdr.TransactionV0(attrs);
-    xtx.operations(this.operations);
+    let txEnvelope;
 
-    const xenv = new xdr.TransactionEnvelope.envelopeTypeTxV0(
-      new xdr.TransactionV0Envelope({ tx: xtx })
-    );
+    if (this.v1) {
+      attrs.sourceAccount = Keypair.fromPublicKey(
+        this.source.accountId()
+      ).xdrAccountId();
+      attrs.ext = new xdr.TransactionExt(0);
+
+      const xtx = new xdr.Transaction(attrs);
+      xtx.operations(this.operations);
+      txEnvelope = new xdr.TransactionEnvelope.envelopeTypeTx(
+        new xdr.TransactionV1Envelope({ tx: xtx })
+      );
+    } else {
+      attrs.sourceAccountEd25519 = Keypair.fromPublicKey(
+        this.source.accountId()
+      )
+        .xdrAccountId()
+        .value();
+      attrs.ext = new xdr.TransactionV0Ext(0);
 
-    const tx = new Transaction(xenv, this.networkPassphrase);
+      const xtx = new xdr.TransactionV0(attrs);
+      xtx.operations(this.operations);
+      txEnvelope = new xdr.TransactionEnvelope.envelopeTypeTxV0(
+        new xdr.TransactionV0Envelope({ tx: xtx })
+      );
+    }
+
+    const tx = new Transaction(txEnvelope, this.networkPassphrase);
 
     this.source.incrementSequenceNumber();
 
     return tx;
   }
+
+  /**
+   * Builds a FeeBumpTransaction
+   * @param {StrKey} feeSource - The account paying for the transaction.
+   * @param {string} baseFee - The max fee willing to pay per operation in inner transaction (**in stroops**). Required.
+   * @param {Transaction} innerTx - The Transaction to be bumped by the fee bump transaction.
+   * @param {string} networkPassphrase - networkPassphrase of the target stellar network (e.g. "Public Global Stellar Network ; September 2015").
+   * @returns {Transaction}
+   * @ignore tell jsdoc to not show this method for now
+   */
+  static buildFeeBumpTransaction(
+    feeSource,
+    baseFee,
+    innerTx,
+    networkPassphrase
+  ) {
+    const innerOps = innerTx.operations.length;
+    const innerBaseFeeRate = new BigNumber(innerTx.fee).div(innerOps);
+    const base = new BigNumber(baseFee);
+
+    // The fee rate for fee bump is at least the fee rate of the inner transaction
+    if (base.lessThan(innerBaseFeeRate)) {
+      throw new Error(
+        `Invalid baseFee, it should be at least ${innerBaseFeeRate} stroops.`
+      );
+    }
+
+    const minBaseFee = new BigNumber(BASE_FEE);
+
+    // The fee rate is at least the minimum fee
+    if (base.lessThan(minBaseFee)) {
+      throw new Error(
+        `Invalid baseFee, it should be at least ${minBaseFee} stroops.`
+      );
+    }
+
+    const fee = base.mul(innerOps + 1);
+
+    const tx = new xdr.FeeBumpTransaction({
+      feeSource: feeSource.xdrAccountId(),
+      fee: xdr.Int64.fromString(fee.toString()),
+      innerTx: xdr.FeeBumpTransactionInnerTx.envelopeTypeTx(
+        innerTx.toEnvelope().value()
```

Comment:
shouldn't we be extracting the v1 arm of innerTx? I think it would also be good to validate that `innerTx` is a v1 transaction (not a v0 tx or a fee bump tx) since the error message we would provide is easier to understand then the one generated by the library crashing

Reply:
We can get the value directly or through the discriminant, the reason I went with the value was to do the validation at the end when converting to an XDR.  I'll update it to explicitly use v1 and validate that the tx is a v1 tx envelope

Reply:
handled in e0d734f